File: samples/CSharp/SourceGenerators/SourceGeneratorSamples/MustacheGenerator.cs
namespace Mustache;

[Generator]
public class MustacheGenerator : IIncrementalGenerator
{
    private const string attributeSource = @"
    [System.AttributeUsage(System.AttributeTargets.Assembly, AllowMultiple=true)]
    internal sealed class MustacheAttribute: System.Attribute
    {
        public string Name { get; }
        public string Template { get; }
        public string Hash { get; }
        public MustacheAttribute(string name, string template, string hash)
            => (Name, Template, Hash) = (name, template, hash);
    }
";

    static string SourceFileFromMustachePath(string name, string template, string hash)
    {
        Func<object, string> tree = HandlebarsDotNet.Handlebars.Compile(template);
        object @object = Newtonsoft.Json.JsonConvert.DeserializeObject(hash);
        string mustacheText = tree(@object);

        StringBuilder sb = new ();
        sb.Append($@"
namespace Mustache {{

    public static partial class Constants {{

        public const string {name} = @""{mustacheText.Replace("\"", "\"\"")}"";
    }}
}}
");
        return sb.ToString();
    }

    public void Initialize(IncrementalGeneratorInitializationContext context)
    {
        context.RegisterPostInitializationOutput(
            ctx => ctx.AddSource("Mustache_MainAttributes__.g.cs", attributeSource));

        var maybeTemplateData = context.SyntaxProvider.CreateSyntaxProvider(
            predicate: (s, t) => s is AttributeSyntax attrib
                                 && attrib.ArgumentList?.Arguments.Count == 3,
            transform: GetTemplateData);

        var templateData = maybeTemplateData
            .Where(t => t is not null)
            .Select((t, _) => t!.Value);

        context.RegisterSourceOutput(templateData,
            (ctx, data) => ctx.AddSource(
                $"{data.name}.g.cs",
                SourceFileFromMustachePath(data.name, data.template, data.hash)));
    }

    internal static (string name, string template, string hash)?
        GetTemplateData(GeneratorSyntaxContext context, CancellationToken cancellationToken)
    {
        var attrib = (AttributeSyntax)context.Node;
        
        if(attrib.ArgumentList is not null
           && context.SemanticModel.GetTypeInfo(attrib).Type?.ToDisplayString() == "MustacheAttribute")
        {
            string name = context.SemanticModel.GetConstantValue(attrib.ArgumentList.Arguments[0].Expression).ToString();
            string template = context.SemanticModel.GetConstantValue(attrib.ArgumentList.Arguments[1].Expression).ToString();
            string hash = context.SemanticModel.GetConstantValue(attrib.ArgumentList.Arguments[2].Expression).ToString();
            return (name, template, hash);
        }
        return null;
    }
}

File: samples/VisualBasic/SourceGenerators/SourceGeneratorSamples/MustacheGenerator.vb
Option Explicit On
Option Infer On
Option Strict On

Imports Microsoft.CodeAnalysis

Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace SourceGeneratorSamples

    <Generator(LanguageNames.VisualBasic)>
    Public Class MustacheGenerator
        Implements ISourceGenerator

        Const attributeSource = "Option Explicit On
Option Strict On
Option Infer On

Namespace Global

    <System.AttributeUsage(System.AttributeTargets.Assembly, AllowMultiple:=True)>
    Friend NotInheritable Class MustacheAttribute
        Inherits System.Attribute

        Public ReadOnly Property Name As String
        Public ReadOnly Property Template As String
        Public ReadOnly Property Hash As String

        Public Sub New(name As String, template As String, hash As String)
            Me.Name = name
            Me.Template = template
            Me.Hash = hash
        End Sub

    End Class

End Namespace
"

        Public Sub Execute(context As GeneratorExecutionContext) Implements ISourceGenerator.Execute
            Dim rx = CType(context.SyntaxContextReceiver, SyntaxReceiver)
            For Each element In rx.TemplateInfo
                Dim source = SourceFileFromMustachePath(element.name, element.template, element.hash)
                context.AddSource($"Mustache{element.name}", source)
            Next
        End Sub

        Private Shared Function SourceFileFromMustachePath(name As String, template As String, hash As String) As String
            Dim tree = HandlebarsDotNet.Handlebars.Compile(template)
            Dim [object] = Newtonsoft.Json.JsonConvert.DeserializeObject(hash)
            Dim mustacheText = tree([object])

            Return $"
Namespace Global.Mustache

    Partial Public Module Constants

        Public Const {name} As String = ""{mustacheText.Replace("""", """""")}""

    End Module

End Namespace"
        End Function

        Public Sub Initialize(context As GeneratorInitializationContext) Implements ISourceGenerator.Initialize
            context.RegisterForPostInitialization(Sub(pi) pi.AddSource("Mustache_MainAttributes__", attributeSource))
            context.RegisterForSyntaxNotifications(Function() New SyntaxReceiver())
        End Sub

        Private Class SyntaxReceiver
            Implements ISyntaxContextReceiver

            Public TemplateInfo As New List(Of (name As String, template As String, hash As String))

            Public Sub OnVisitSyntaxNode(context As GeneratorSyntaxContext) Implements ISyntaxContextReceiver.OnVisitSyntaxNode
                If context.Node.Kind() = SyntaxKind.Attribute Then
                    Dim attrib = CType(context.Node, AttributeSyntax)
                    If attrib.ArgumentList?.Arguments.Count = 3 AndAlso
                       context.SemanticModel.GetTypeInfo(attrib).Type?.ToDisplayString() = "MustacheAttribute" Then
                        Dim name = context.SemanticModel.GetConstantValue(attrib.ArgumentList.Arguments(0).GetExpression).ToString()
                        Dim template = context.SemanticModel.GetConstantValue(attrib.ArgumentList.Arguments(1).GetExpression).ToString()
                        Dim hash = context.SemanticModel.GetConstantValue(attrib.ArgumentList.Arguments(2).GetExpression).ToString()
                        TemplateInfo.Add((name, template, hash))
                    End If
                End If
            End Sub
        End Class

    End Class

End Namespace
